test/lit/passes/optimize-instructions-gc.wast

```diff
@@ -936,4 +935,126 @@
       )
     )
   )
+
+  ;; CHECK:      (func $ref-cast-squared (param $x eqref)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:    (rtt.canon $struct)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $ref-cast-squared (param $x eqref)
+    ;; Identical ref.casts can be folded together.
+    (drop
+      (ref.cast
+        (ref.cast
+          (local.get $x)
+          (rtt.canon $struct)
+        )
+        (rtt.canon $struct)
+      )
+    )
+  )
+  ;; CHECK:      (func $ref-cast-squared-fallthrough (param $x eqref)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.tee $x
+  ;; CHECK-NEXT:    (ref.cast
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:     (rtt.canon $struct)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $ref-cast-squared-fallthrough (param $x eqref)
+    ;; A fallthrough in the middle does not prevent this optimization.
+    (drop
+      (ref.cast
+        (local.tee $x
+          (ref.cast
+            (local.get $x)
+            (rtt.canon $struct)
+          )
+        )
+        (rtt.canon $struct)
+      )
+    )
+  )
+  ;; CHECK:      (func $ref-cast-cubed (param $x eqref)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast
+  ;; CHECK-NEXT:    (local.get $x)
+  ;; CHECK-NEXT:    (rtt.canon $struct)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $ref-cast-cubed (param $x eqref)
+    ;; Three and more also work.
+    (drop
+      (ref.cast
+        (ref.cast
+          (ref.cast
+            (local.get $x)
+            (rtt.canon $struct)
+          )
+          (rtt.canon $struct)
+        )
+        (rtt.canon $struct)
+      )
+    )
+  )
+  ;; CHECK:      (func $ref-cast-squared-no-different (param $x eqref)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.cast
+  ;; CHECK-NEXT:    (ref.cast
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:     (rtt.canon $empty)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (rtt.canon $struct)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $ref-cast-squared-no-different (param $x eqref)
```

Comment:
I would remove "no-" from these names; it makes it sound like the test will contain no difference or no effects, but actually you mean the opposite.

Reply:
Good point, will update.